### PR TITLE
log tests

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This enhances test coverage and follows up on the contextual logging enabling.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/component-base/logs/json/klog_test.go [hunk 1]
```diff
@@ -239,7 +240,9 @@ func TestKlogIntegration(t *testing.T) {
 		t.Run(tc.name, func(t *testing.T) {
 			var buffer bytes.Buffer
 			writer := zapcore.AddSync(&buffer)
-			logger, _ := NewJSONLogger(100, writer, nil, nil)
+			// This level is higher enough to enable all log messages from this test.
```
Comment (logicalhan) [inline staging/src/k8s.io/component-base/logs/json/klog_test.go:243]: nit: 'high enough'

**Review by logicalhan — COMMENTED** [review]
Can you paste the contents of the benchmarks into the PR itself? I'm quite curious what the results look like.

**Comment by pohly** [comment]
Here are the results of `go test -bench -count=10 ./test/integration/logs/benchmark/contextual-logging`, summarized by benchstat:
```
name                                         time/op
Recursion/10-36                              20.3ns ± 3%
Recursion/100-36                              157ns ± 0%
Recursion/1000-36                            1.62µs ± 0%
Recursion/10000-36                           17.1µs ± 3%
Recursion/100000-36                           457µs ± 2%
RecursionWithLogger/10-36                    17.7ns ± 0%
RecursionWithLogger/100-36                    212ns ± 0%
RecursionWithLogger/1000-36                  1.97µs ± 1%
RecursionWithLogger/10000-36                 22.3µs ± 1%
RecursionWithLogger/100000-36                 350µs ± 3%
RecursionWithContext/10-36                   17.7ns ± 0%
RecursionWithContext/100-36                   210ns ± 3%
RecursionWithContext/1000-36                 1.91µs ± 1%
RecursionWithContext/10000-36                23.1µs ± 1%
RecursionWithContext/100000-36                314µs ± 2%
RecursionWithLoggerAndLog/10-36              18.1ns ± 0%
RecursionWithLoggerAndLog/100-36              172ns ± 0%
RecursionWithLoggerAndLog/1000-36            2.11µs ± 2%
RecursionWithLoggerAndLog/10000-36           20.9µs ± 8%
RecursionWithLoggerAndLog/100000-36           516µs ± 2%
RecursionWithContextAndLog/10-36             28.3ns ± 0%
RecursionWithContextAndLog/100-36             198ns ± 1%
RecursionWithContextAndLog/1000-36           2.35µs ± 1%
RecursionWithContextAndLog/10000-36          22.9µs ± 5%
RecursionWithContextAndLog/100000-36          513µs ± 3%
NestedContextWithTimeouts/1-36               1.00µs ± 2%
NestedContextWithTimeouts/10-36              12.0µs ± 2%
NestedContextWithTimeouts/100-36              142µs ± 1%
NestedContextWithTimeouts/1000-36            3.81ms ± 7%
NestedContextWithTimeouts/10000-36            203ms ± 2%
NestedContextWithTimeoutsAndLookup/1-36      1.05µs ± 5%
NestedContextWithTimeoutsAndLookup/10-36     12.5µs ± 4%
NestedContextWithTimeoutsAndLookup/100-36     150µs ± 3%
NestedContextWithTimeoutsAndLookup/1000-36   3.91ms ± 4%
NestedContextWithTimeoutsAndLookup/10000-36   204ms ± 3%
LookupWithTimeouts/1-36                      7.72ns ± 0%
LookupWithTimeouts/10-36                     22.0ns ± 1%
LookupWithTimeouts/100-36                     179ns ± 0%
LookupWithTimeouts/1000-36                   3.02µs ± 1%
LookupWithTimeouts/10000-36                  30.4µs ± 1%
LookupWithValues/1-36                        7.78ns ± 0%
LookupWithValues/10-36                       20.1ns ± 1%
LookupWithValues/100-36                       156ns ± 2%
LookupWithValues/1000-36                     2.08µs ± 1%
LookupWithValues/10000-36                    21.0µs ± 0%
```

Looking at the benchmark again, I had noticed that the `FromContextOrDiscard` at the leaf of  `RecursionWithContext` may have handicapped that variant, so I removed logging from that and `RecursionWithLogger` benchmarks to focus exclusively on argument passing. I also updated the lookup benchmarks.

`Recursion` is the baseline. It simulates functions up to different depths (10, 100, etc.).

`RecursionWithLogger` and `RecursionWithContext` simulate adding a `Logger` resp. `context` parameter *without* logging. Comparing the two, there's no big difference anymore without the logging. But passing an additional parameter without ever using it is not realistic. The `*AndLog` variants add that, and here passing a `Logger` has the advantage that `RecursionWithContext` isn't necessary. That is where the performance advantage came from that I saw earlier and mention in https://github.com/kubernetes/community/pull/6560.

`NestedContextWithTimeouts` shows how expensive it is to create contexts with `WithTimeout`. Doing one lookup at the end (`NestedContextWithTimeoutsAndLookup`) doesn't change that much.

The cost for that lookup primarily depends on how deeply nested the contexts are and increases linearly with the nesting depth because it needs to search backwards in a linked list. There's no big difference what function creates the new context (`LookupWithTimeouts` vs. `LookupWithValues`).

File: staging/src/k8s.io/component-base/logs/json/klog_test.go (hunk 1 above)
Comment (pohly) [reply staging/src/k8s.io/component-base/logs/json/klog_test.go]: Fixed

**Comment by pohly** [comment]
@logicalhan: can you perhaps take another look?

**Pull request merged**